lib/collection/src/collection/shard_transfer.rs

```diff
@@ -418,7 +420,15 @@ impl Collection {
             }
 
             if replica_set.is_dummy().await {
+                // Check if shard was dirty before init_empty_local_shard
+                let was_dirty = replica_set.is_dirty().await;
+                // TODO: If dirty, still try to load the shard and init empty shard only if it's not recoverable?
                 replica_set.init_empty_local_shard().await?;
+
+                if was_dirty {
+                    let shard_flag = shard_initializing_flag_path(&collection_path, shard_id);
+                    tokio::fs::remove_file(&shard_flag).await?;
+                }
             }
```

Comment:
Yes, we should initialize an empty shard here as cannot do our recovery process on top of a corrupt shard, and so we must empty it here.

```suggestion
            if replica_set.is_dummy().await {
                // If shard was dirty, remove initializing flag after initializing empty
                let was_dirty = replica_set.is_dirty().await;
                replica_set.init_empty_local_shard().await?;
                if was_dirty {
                    let shard_flag = shard_initializing_flag_path(&collection_path, shard_id);
                    tokio::fs::remove_file(&shard_flag).await?;
                }
            }
```

It's fine to empty it here, because at this point we're sure that there's some other replica that still has our data. We therefore have no data loss.

Reply:
> because at this point we're sure that there's some other replica that still has our data. We therefore have no data loss

Because there was a shard initializing flag? Hmm, makes sense.